doc/neps/nep-0035-array-creation-dispatch-with-array-function.rst

```diff
@@ -8,16 +8,176 @@ NEP 35 — Array Creation Dispatching With __array_function__
+array or any other array from a library that adopts ``__array_function__``.
+This allows writing code that is agnostic to the implementation library, thus
+users can write their code once and still be able to use different array
+implementations according to their needs.
+
+Obviously, having a protocol in-place is useful if the arrays are created
+elsewhere and let NumPy handle them. But still these arrays have to be started
+in their native library and brought back. Instead if it was possible to create
+these objects through NumPy API then there would be an almost complete
+experience, all using NumPy syntax. For example, say we have some CuPy array
+``cp_arr`` , and want a similar CuPy array with identity matrix. We could still
+write the following:
+
+.. code:: python
+     x = cupy.identity(3)
+
+Instead, the better way would be using to only use the NumPy API, this could now
+be achieved with:
+
+.. code:: python
+    x = np.identity(3, like=cp_arr)
+
+As if by magic, ``x`` will also be a CuPy array, as NumPy was capable to infer
+that from the type of ``cp_arr``. Note that this last step would not be possible
+without ``like=``, as it would be impossible for the NumPy to know the user
+expects a CuPy array based only on the integer input.
+
+The new ``like=`` keyword proposed is solely intended to identify the downstream
+library where to dispatch and the object is used only as reference, meaning that
+no modifications, copies or processing will be performed on that object.
+
+We expect that this functionality will be mostly useful to library developers,
+allowing them to create new arrays for internal usage based on arrays passed
+by the user, preventing unnecessary creation of NumPy arrays that will
+ultimately lead to an additional conversion into a downstream array type.
+
+Support for Python 2.7 has been dropped since NumPy 1.17, therefore we make use
+of the keyword-only argument standard described in PEP-3102 [2]_ to implement
+``like=``, thus preventing it from being passed by position.
+
+.. _neps.like-kwarg.usage-and-impact:
+
+Usage and Impact
+----------------
+
+To understand the intended use for ``like=``, and before we move to more complex
+cases, consider the following illustrative example consisting only of NumPy and
+CuPy arrays:
+
+.. code:: python
+
+    import numpy as np
+    import cupy
+
+    def my_pad(arr, padding):
+        padding = np.array(padding, like=arr)
+        return np.concatenate((padding, arr, padding))
+
+    my_pad(np.arange(5), [-1, -1])    # Returns np.ndarray
+    my_pad(cupy.arange(5), [-1, -1])  # Returns cupy.core.core.ndarray
+
+Note in the ``my_pad`` function above how ``arr`` is used as a reference to
+dictate what array type padding should have, before concatenating the arrays to
+produce the result. On the other hand, if ``like=`` wasn't used, the NumPy case
+case would still work, but CuPy wouldn't allow this kind of automatic
```

Comment:
```suggestion
would still work, but CuPy wouldn't allow this kind of automatic
```

Reply:
What happens in this example if I call `my_pad([1, 2, 3], [-1, 1])`?

Reply:
@mattip: Good question. I'd prefer the answer to be "coercion to `np.array`" I think.

Reply:
> What happens in this example if I call my_pad([1, 2, 3], [-1, 1])?

Same as you would see without `like=arr`:

```python
In [1]: import numpy as np
   ...:
   ...: def my_pad(arr, padding):
   ...:     padding = np.array(padding, like=arr)
   ...:     return np.concatenate((padding, arr, padding))
   ...:
   ...: a = my_pad([1, 2, 3], [-1, -1])
   ...: print(type(a))
   ...: print(a)
   ...:
<class 'numpy.ndarray'>
[-1 -1  1  2  3 -1 -1]
```

Reply:
I meant that in this case it would call `np.array([-1, -1], like=[1, 2, 3])`, and `list` doesn't have an `__array_function__`. Would it raise an error? I guess a more complete example would call `arr = np.arraylike(arr)` in the first line, so never mind.

Reply:
It wouldn't raise an error, it's expected to work as other functions implementing `__array_function__`, where dispatching on a `list` would just fallback to NumPy's default behavior.